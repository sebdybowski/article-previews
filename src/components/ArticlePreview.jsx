import React from 'react';
import PropTypes from 'prop-types';

export const ArticlePreview = ({ article }) => <article className="card mb-3">
    <div className="row no-gutters">
        <div className="col-md-4">
            <img src={article.image} className="card-img" alt={`${article.title} image`} />
        </div>
        <div className="col-md-8">
            <div className="card-body">
                <div className="row">
                    <div className="col-8">
                        <h5 className="card-title">{article.title}</h5>
                    </div>
                    <div className="col-4">
                        <small className="text-muted float-right">{article.date}</small>
                    </div>
                </div>
                <p className="card-text">{article.preamble}</p>
            </div>
        </div>
    </div>
</article>;

ArticlePreview.propTypes = {
    article: PropTypes.shape({
        category: PropTypes.string,
        date: PropTypes.string,
        id: PropTypes.number,
        image: PropTypes.string,
        preamble: PropTypes.string,
        title: PropTypes.string,
    }),
};

ArticlePreview.defaultProps = {
    article: [],
};
