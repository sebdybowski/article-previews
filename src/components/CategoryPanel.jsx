import React from 'react';
import PropTypes from 'prop-types';

export const CategoryPanel = () => <ul className="nav flex-column">
    <li className="nav-item">
        <a className="nav-link active" href="#">Sports</a>
    </li>
    <li className="nav-item">
        <a className="nav-link" href="#">Fashion</a>
    </li>
</ul>;