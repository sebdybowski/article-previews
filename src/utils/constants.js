export const ENDPOINT_TYPE = {
    SPORTS: 'sports',
    FASHION: 'fashion',
};

export const CATEGORY_TYPE = ['sport', 'fashion'];

export const INITIAL_STATE = {
    articles: [],
    selectedCategories: CATEGORY_TYPE,
};