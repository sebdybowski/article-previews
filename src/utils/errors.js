import _ from 'lodash';

export const resolveErrorCategory = errorData => _.chain(errorData)
    .get('response.config.url')
    .split('/')
    .last()
    .value();