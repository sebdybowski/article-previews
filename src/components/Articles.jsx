import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { ArticlePreview } from './ArticlePreview';

export const Articles = ({ articles }) => _.map(articles, article =>
    <ArticlePreview article={article} key={article.id}/>);

Articles.propTypes = {
    articles: PropTypes.arrayOf(PropTypes.shape({
        category: PropTypes.string,
        date: PropTypes.string,
        dateSortable: PropTypes.number,
        id: PropTypes.number,
        image: PropTypes.string,
        preamble: PropTypes.string,
        title: PropTypes.string,
    })),
};