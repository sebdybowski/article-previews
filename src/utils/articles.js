import _ from 'lodash';
import {MONTH_MAP, SORT_ORDER} from "./constants";

export const resolveSortableDate = article => _.chain(article)
    .get('date')
    .split(' ')
    .map((value, index) => {
        if (index === 0 && value.length <= 2 ) return `0${value}`;
        if (index === 1) return _.get(MONTH_MAP, value);
        return value;
    })
    .reverse()
    .join('')
    .replace('.', '')
    .toNumber()
    .value();

export const resolveArticlesList = (articles, selectedCategories, sortOrder) => _.chain(articles)
    .filter(article => _.includes(selectedCategories, _.get(article, 'category')))
    .orderBy(['dateSortable'], [sortOrder])
    .value();
