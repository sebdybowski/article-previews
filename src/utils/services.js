import axios from 'axios';
import _ from 'lodash';
import { ENDPOINT_TYPE } from './constants';

const { SPORTS, FASHION } = ENDPOINT_TYPE;

export const fetchArticle = type => axios.get(`http://localhost:6010/articles/${type}`)
    .then(response => _.get(response, 'data.articles'))
    .catch(error => {
        console.log(error);
        return error;
    });

export const fetchAllArticles = () => Promise.all([fetchArticle(SPORTS), fetchArticle(FASHION)])
    .then(response => _.flatten(response))
    .catch(error => console.log(error));