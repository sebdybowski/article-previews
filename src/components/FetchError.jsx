import React from 'react';
import PropTypes from 'prop-types';
import { resolveErrorCategory } from '../utils/errors';

export const FetchError = ({ errorData }) =>
    <div className="alert alert-danger" role="alert">
        <p>
            { 'We are sorry, but requests from category ' }
            <span className="font-weight-bold">{resolveErrorCategory(errorData)}</span>
            { ' were not loaded due to ' }
            <span className="font-weight-bold">{errorData.response.status}</span>
            { ' error. This bug is temporary, please: ' }
        </p>
        <button onClick={() => window.location.reload()} type="button" className="btn btn-primary btn-block">Reload the page</button>
    </div>;

FetchError.propTypes = {
    errorData: PropTypes.object,
};