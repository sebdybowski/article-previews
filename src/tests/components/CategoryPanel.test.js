import React from 'react';
import { shallow, render } from 'enzyme';
import { CategoryPanel } from '../../components/CategoryPanel';
import _ from 'lodash';
import { CATEGORY_TYPE } from '../../utils/constants';

const mockProps = {
    selectedValues: CATEGORY_TYPE,
    onValueChange: _.noop,
};

describe('CategoryPanel', () => {
    it('should render correctly with default props', () => {
        const component = shallow(<CategoryPanel {...mockProps} />);
        expect(component).toMatchSnapshot();
    });
    it('should render correctly first value selected', () => {
        const component = shallow(<CategoryPanel {...mockProps} selectedValues={[CATEGORY_TYPE[0]]} />);
        expect(component).toMatchSnapshot();
    });
    it('should render correctly second value selected', () => {
        const component = shallow(<CategoryPanel {...mockProps} selectedValues={[CATEGORY_TYPE[1]]} />);
        expect(component).toMatchSnapshot();
    });
    it('should render correctly no value selected', () => {
        const component = shallow(<CategoryPanel {...mockProps} selectedValues={null} />);
        expect(component).toMatchSnapshot();
    });
});