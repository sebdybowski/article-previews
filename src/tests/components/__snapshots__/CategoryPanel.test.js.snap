// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryPanel should render correctly first value selected 1`] = `
<ul
  className="nav flex-column"
>
  <h5
    className="mb-3"
  >
    Data sources
  </h5>
  <li
    className="nav-item mb-2"
    key="sport"
  >
    <Checkbox
      label="sport"
      onValueChange={[Function]}
      value={true}
    />
  </li>
  <li
    className="nav-item mb-2"
    key="fashion"
  >
    <Checkbox
      label="fashion"
      onValueChange={[Function]}
      value={false}
    />
  </li>
</ul>
`;

exports[`CategoryPanel should render correctly no value selected 1`] = `
<ul
  className="nav flex-column"
>
  <h5
    className="mb-3"
  >
    Data sources
  </h5>
  <li
    className="nav-item mb-2"
    key="sport"
  >
    <Checkbox
      label="sport"
      onValueChange={[Function]}
      value={false}
    />
  </li>
  <li
    className="nav-item mb-2"
    key="fashion"
  >
    <Checkbox
      label="fashion"
      onValueChange={[Function]}
      value={false}
    />
  </li>
</ul>
`;

exports[`CategoryPanel should render correctly second value selected 1`] = `
<ul
  className="nav flex-column"
>
  <h5
    className="mb-3"
  >
    Data sources
  </h5>
  <li
    className="nav-item mb-2"
    key="sport"
  >
    <Checkbox
      label="sport"
      onValueChange={[Function]}
      value={false}
    />
  </li>
  <li
    className="nav-item mb-2"
    key="fashion"
  >
    <Checkbox
      label="fashion"
      onValueChange={[Function]}
      value={true}
    />
  </li>
</ul>
`;

exports[`CategoryPanel should render correctly with default props 1`] = `
<ul
  className="nav flex-column"
>
  <h5
    className="mb-3"
  >
    Data sources
  </h5>
  <li
    className="nav-item mb-2"
    key="sport"
  >
    <Checkbox
      label="sport"
      onValueChange={[Function]}
      value={true}
    />
  </li>
  <li
    className="nav-item mb-2"
    key="fashion"
  >
    <Checkbox
      label="fashion"
      onValueChange={[Function]}
      value={true}
    />
  </li>
</ul>
`;
